source/common/ssl/context_config_impl.cc

```diff
@@ -34,6 +34,61 @@ const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES =
 
 const std::string ContextConfigImpl::DEFAULT_ECDH_CURVES = "X25519:P-256";
 
+static std::vector<bssl::UniquePtr<X509_CRL>>
+loadCRL(const envoy::api::v2::CertificateValidationContext& vctx) {
+  std::vector<bssl::UniquePtr<X509_CRL>> crls;
+
+  if (!vctx.has_crl()) {
+    return crls;
+  }
+
+  // We want to load multiple CRLs from the same file, so we need to use
+  // OpenSSL's STACK abstraction. We read from the input file or inline data,
+  // and load into the stack of X509_INFOs.
+  bssl::UniquePtr<STACK_OF(X509_INFO)> xis(nullptr);
+
+  auto crl_conf = vctx.crl();
+  if (crl_conf.filename().length()) {
+    std::unique_ptr<FILE, decltype(&fclose)> fp(fopen(crl_conf.filename().c_str(), "r"), &fclose);
+
+    if (!fp.get()) {
+      throw EnvoyException(
+          fmt::format("Failed to open CRL file '{}'", crl_conf.filename().c_str()));
+    }
+
+    xis.reset(PEM_X509_INFO_read(fp.get(), nullptr, nullptr, nullptr));
+  } else {
+    std::unique_ptr<BIO, decltype(&BIO_free_all)> bio(
+        BIO_new_mem_buf(crl_conf.inline_().c_str(), crl_conf.inline_().length()), BIO_free_all);
+
+    if (!bio.get()) {
+      throw EnvoyException(fmt::format("Failed to load inline CRL"));
+    }
+
+    xis.reset(PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));
+  }
+
+  if (!xis) {
+    throw EnvoyException(fmt::format("Failed to parse CRL"));
+  }
+
+  // Iterate through every item in the stack, adding all found CRLs to our
+  // return array.
+  for (unsigned int i = 0; i < sk_X509_INFO_num(xis.get()); i++) {
```

Comment:
You might want to look at similar code in https://github.com/envoyproxy/envoy/pull/2248, and use the list iterator from `bssl::UniquePtr`.

Reply:
Ooh, that's good to know - I didn't realize there was an iterator on `STACK`s.  Also, regarding the similar code: as I read it, #2248 will load CRLs from the file passed as a CA cert; I don't want to duplicate work, but $WORK definitely needs to be able to load CRLs from a different file from CAs (we essentially fetch CRLs on a cronjob, whereas the CA certs are placed with config management).  What are your thoughts on reducing duplication here?

Reply:
Yeah, CRLs will be loaded from `trusted_ca` as well, since that's what the filesystem-based API is currently doing, and I didn't want to change the behavior and suddenly start ignoring CRLs from the `trusted_ca` file... Having said that, I don't think that's a common use case and handling CRLs via separate field is definitely OK.